src/node_url_pattern.cc

```diff
@@ -0,0 +1,571 @@
+  if (init.has_value()) {
+    url_pattern = ada::parse_url_pattern(
+        *init,
+        base_url ? &base_url_view : nullptr,
+        options.has_value() ? &options.value() : nullptr);
+  } else {
+    url_pattern = ada::parse_url_pattern(
+        *input,
+        base_url ? &base_url_view : nullptr,
+        options.has_value() ? &options.value() : nullptr);
+  }
+
+  if (!url_pattern) {
+    env->ThrowTypeError("Failed to construct URLPattern");
+  }
+
+  new URLPattern(env, args.This(), std::move(*url_pattern));
+}
+
+ada::url_pattern_init URLPattern::URLPatternInit::FromJsObject(
+    Environment* env, Local<Object> obj) {
+  ada::url_pattern_init init{};
+  Local<String> components[] = {
+      env->protocol_string(),
+      env->username_string(),
+      env->password_string(),
+      env->hostname_string(),
+      env->port_string(),
+      env->pathname_string(),
+      env->search_string(),
+      env->hash_string(),
+      env->base_url_string(),
+  };
+  Local<Value> value;
+  auto isolate = env->isolate();
+  auto set_parameter = [&](Local<String> component,
+                           std::string_view utf8_value) {
+    // TODO(anonrig): Optimization opportunity.
+    // Get rid of calling env->xxx_string() for every component match.
+    if (component == env->protocol_string()) {
+      init.protocol = std::string(utf8_value);
+    } else if (component == env->username_string()) {
+      init.username = std::string(utf8_value);
+    } else if (component == env->password_string()) {
+      init.password = std::string(utf8_value);
+    } else if (component == env->hostname_string()) {
+      init.hostname = std::string(utf8_value);
+    } else if (component == env->port_string()) {
+      init.port = std::string(utf8_value);
+    } else if (component == env->pathname_string()) {
+      init.pathname = std::string(utf8_value);
+    } else if (component == env->search_string()) {
+      init.search = std::string(utf8_value);
+    } else if (component == env->hash_string()) {
+      init.hash = std::string(utf8_value);
+    } else if (component == env->base_url_string()) {
+      init.base_url = std::string(utf8_value);
+    }
+  };
+  for (const auto& component : components) {
+    if (obj->Get(env->context(), component).ToLocal(&value)) {
+      if (value->IsString()) {
+        Utf8Value utf8_value(isolate, value);
+        set_parameter(component, utf8_value.ToStringView());
+      }
+    }
+  }
+  return init;
+}
+
+Local<Object> URLPattern::URLPatternComponentResult::ToJSObject(
+    Environment* env, const ada::url_pattern_component_result& result) {
+  auto isolate = env->isolate();
+  Local<Name> names[] = {env->input_string(), env->groups_string()};
+  Local<Value> values[] = {
+      ToV8Value(env->context(), result.input).ToLocalChecked(),
+      ToV8Value(env->context(), result.groups).ToLocalChecked(),
+  };
+  return Object::New(isolate, Null(isolate), names, values, 2);
```

Comment:
```suggestion
  return Object::New(isolate, Null(isolate), names, values, arraysize(names));
```

here and the other places you`re using that pattern